File: books/clrs/ch1/insertion_sort.py
#!/usr/bin/env python3
import operator
from typing import List

def insertion_sort(a: List, compare=operator.gt) -> List:
    """Insertion sort algorithm"""
    for j in range(1, len(a)):
        key = a[j]

        i = j - 1
        while i >= 0 and compare(a[i], key):
            a[i + 1] = a[i]
            i = i - 1
        a[i + 1] = key

    return a

if __name__ == "__main__":
    ints = [-2, 99, 0, -743, 2, 3, 4]

    print(
        insertion_sort(ints),
        insertion_sort(ints, operator.lt),
        sep='\n'
    )

File: books/clrs/ch1/insertion_sort.c
#include <stdio.h>

void insertion_sort(int* A, size_t l)
{
    for (int j = 1; j < l; j++) {
        int key = A[j];

        int i;
        for (i = j - 1; i >= 0 && A[i] >= key; i--) {
            A[i + 1] = A[i];
        }

        A[i + 1] = key;
    }
}

int main(void)
{
    int ints[] = { -2, 99, 0, -743, 2, 3, 4 };
    int size = sizeof(ints) / sizeof(ints[0]);

    insertion_sort(ints, size);
 
    for (int i = 0; i < size; i++)
        printf("%d ", ints[i]);
 
    printf("\n");
 
    return 0;
}